torchx/examples/mnist.exs

```diff
@@ -0,0 +1,189 @@
+defmodule Torchx.MNIST do
+  import Nx.Defn
+
+  @batch_size 50
+  @step 0.02
+  @epochs 10
+
+  defn init_random_params do
+    w1 = Nx.random_normal({784, 128}, 0.0, 0.1, names: [:input, :layer])
+    b1 = Nx.random_normal({128}, 0.0, 0.1, names: [:layer])
+    w2 = Nx.random_normal({128, 10}, 0.0, 0.1, names: [:layer, :output])
+    b2 = Nx.random_normal({10}, 0.0, 0.1, names: [:output])
+    {w1, b1, w2, b2}
+  end
+
+  defn softmax(logits) do
+    exp = Nx.exp(logits)
+    Nx.divide(exp, Nx.sum(exp, axes: [:output], keep_axes: true))
+  end
+
+  defn predict({w1, b1, w2, b2}, batch) do
+    batch
+    |> Nx.dot(w1)
+    |> Nx.add(b1)
+    |> Nx.logistic()
+    |> Nx.dot(w2)
+    |> Nx.add(b2)
+    |> softmax()
+  end
+
+  defn update(
+         {w1, b1, w2, b2} = _params,
+         batch_images,
+         batch_labels,
+         avg_loss,
+         avg_accuracy,
+         total
+       ) do
+    z1 = Nx.dot(batch_images, w1) |> Nx.add(b1)
+    a1 = Nx.logistic(z1)
+    z2 = Nx.dot(a1, w2) |> Nx.add(b2)
+    preds = softmax(z2)
+
+    batch_loss =
+      Nx.sum(Nx.mean(Nx.multiply(Nx.log(preds), batch_labels), axes: [:output])) |> Nx.negate()
+
+    batch_accuracy =
+      Nx.mean(
+        Nx.equal(
+          Nx.argmax(Nx.as_type(batch_labels, {:s, 8}), axis: :output),
+          Nx.argmax(preds, axis: :output)
+        )
+        |> Nx.as_type({:s, 8})
+      )
+
+    total = Nx.tensor(total)
+
+    avg_loss = Nx.add(avg_loss, Nx.divide(batch_loss, total))
+    avg_accuracy = Nx.add(avg_accuracy, Nx.divide(batch_accuracy, total))
+
+    grad_z2 = Nx.subtract(preds, batch_labels) |> Nx.transpose()
+
+    batch_size = Nx.tensor(@batch_size)
+
+    grad_w2 = Nx.divide(Nx.dot(grad_z2, a1), batch_size) |> Nx.transpose()
+    grad_b2 = Nx.mean(Nx.transpose(grad_z2), axes: [:output], keep_axes: true)
+
+    grad_a1 = Nx.dot(w2, grad_z2) |> Nx.transpose()
+    grad_z1 = Nx.multiply(grad_a1, a1) |> Nx.multiply(Nx.subtract(Nx.tensor(1.0), a1))
+
+    grad_w1 = Nx.divide(Nx.dot(Nx.transpose(grad_z1), batch_images), batch_size) |> Nx.transpose()
+    grad_b1 = Nx.mean(grad_z1, axes: [1], keep_axes: true)
+
+    step = Nx.tensor(@step)
+
+    {{
+       Nx.subtract(w1, Nx.multiply(grad_w1, step)),
+       Nx.subtract(b1, Nx.multiply(grad_b1, step)),
+       Nx.subtract(w2, Nx.multiply(grad_w2, step)),
+       Nx.subtract(b2, Nx.multiply(grad_b2, step))
+     }, avg_loss, avg_accuracy}
+  end
```

Comment:
You implemented the grad by hand but now you can rollback and use `grad` as usual! My suggestion is to copy the example from EXLA again and mostly use it as is, simply removing the `@defn_compiler` directives. :)

Reply:
We need to update the regression example too!